text/0000-project-ffi-unwind.md

```diff
@@ -0,0 +1,211 @@
+
+* A **charter** defining the project's scope
+* A **liaison** with an official Rust team (who may or may not also be a shepherd)
+* A small number of **shepherds**, who are responsible for summarizing
+  conversations and keeping the lang team abreast of interesting developments.
+* A GitHub repository hosted under the `rust-lang` organization containing the
+  charter and instructions for how community members can monitor the group's
+  progress and/or participate.
+
+[This blog post][shepherds-3.0] explains in detail the role of the
+shepherds.
+
+## Project group roadmap and RFCs
+
+The first step of the project group is to define a **roadmap** indicating the
+planned sequence in which it will design and propose particular behaviors and
+features.  Once the project group feels it has completed work on some item in
+the roadmap, that item will be submitted as an RFC or FCP for review by the lang team and the community at large.
+
+## Stabilizing unspecified "TBD" behavior
+[stabilizing-tbd]: stabilizing-unspecified-tbd-behavior
+
+We would like to be able to provide features in stable Rust without a full
+formal specification but with an informal statement of intent regarding the
+behavior. These features would be considered "unspecified behavior" (rather
+than "undefined behavior"), and their behavior would therefore be subject to
+some change, but such a change would be guaranteed not to violate the project
+group's stated intent for the feature.
+
+Internally, a project group should refer to such unspecified behavior as "TBD",
+to indicate that refining the specification of the behavior is within the scope
+of the project. Outside of the context of the project group, such as in general
+reference material for the language, the behavioral intent as determined by the
+project group should be described, but will not imply any special status
+compared to other instances of unspecified behavior in the language.
+
+## Details of the FFI-unwind project group
+
+[Repository][ffi-unwind project]
+
+Initial shepherds:
+
+* [acfoltzer (Adam)](https://github.com/acfoltzer)
+* [batmanaod (Kyle)](https://github.com/batmanaod)
+
+Lang team liaisons:
+
+* [nikmoatsakis (Niko)](https://github.com/nikmoatsakis)
+* [joshtriplett (Josh)](https://github.com/joshtriplett)
+
+### Charter
+[charter]: #charter
+
+The FFI-unwind project group has the following initial scope:
+
+* to define the details of the "C unwind" ABI on major Tier 1 platforms
+* in particular, to define with sufficient detail to enable the use cases
+  described in the Motivation section of this RFC
+  
+Certain elements are definitively out of scope:
+
+* The group does not intend to consider mechanisms to enable "interop"
+  between Rust panics and exceptions from other languges. For example,
+  we do not intend to permit Rust code to catch C++ exceptions, though
+  we will have to consider what happens when a C++ exception unwinds
+  past a `catch_unwind` boundary.
```

Comment:
There are a couple of contradictions in this section.

First, the Motivation only speaks about "unwinding" out of `extern "C"` functions in general, and as a general problem, that includes dealing with foreign exceptions (e.g. due to cancellation points on Linux). 

Second, this paragraph suggests that foreign exceptions will be ignored while simultaneously saying that they will be considered. 

I think this section should clarify whether the intent is for the project group to provide features that will work with foreign exceptions or not at some point. While it is fair to say that supporting foreign exceptions is not part of an MVP, whether they are factored in the initial design or not might be the difference between `"C unwind"` supporting them in the future, or the language having to add a different language feature to support them (e.g. another `"C platform-unwind" ABI).

Reply:
> First, the Motivation only speaks about "unwinding" out of `extern "C"` functions in general, and as a general problem, that includes dealing with foreign exceptions (e.g. due to cancellation points on Linux).

I'm not sure that's a "contradiction"; it just means that some related features that could be helpful are nevertheless out of scope for _this_ group. However, given Amanieu's PR, maybe it's not a great example of something that will _never_ be in scope.

Reply:
> Second, this paragraph suggests that foreign exceptions will be ignored while simultaneously saying that they will be considered.

I don't know why you would say "foreign exceptions will be ignored"; this section is just saying that full exception-interop (e.g. defining how to interact with foreign exception objects) is not in scope.

> I think this section should clarify whether the intent is for the project group to provide features that will work with foreign exceptions or not at some point.

Well, as it stands, it does clarify that, by saying that the group won't provide such features. But if you're saying that this RFC _must_ specify whether or not these features will ever be in scope for the group, then I disagree; I don't see what benefit there would be in saying "that's in scope" or "that's out of scope" before the group has proposed a single technical RFC.

Reply:
Maybe this section should explain in more detail what precisely is in scope and what precisely is not.

> I don't know why you would say "foreign exceptions will be ignored";

That's what this sentence says to me: "The group does not intend to consider mechanisms to enable "interop" between Rust panics and exceptions from other languge". I read that as this working group will not deal with how exceptions from other languages, called "foreign exceptions", interoperate with Rust. Yet the next sentences say that in some cases, this working group will deal with precisely that. 

If that isn't what this sentence tries to convey, what is it? Maybe it could be worded somehow else to prevent me from misunderstanding it in that way ?

Reply:
I think this is another case where it's useful to define expectations. Most important seems to be saying what we DO plan to do. It's also useful to say what we don't *expect* to do, but we can leave room for future extension. I guess here I would say something like:

We expect to define how foreign exceptions interact with `catch_unwind`. We do not currently expect to add new mechanisms for interacting with or throwing foreign exceptions. We will revisit possible extensions when the initial goals have been met.